man/omorfi-vislcg.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.46.6.
.TH OMORFI-VISLCG "1" "June 2018" "OMORFI" "User Commands"
.SH NAME
omorfi-vislcg \- Analyse Finnish into VISL CG 3 format using omorfi
.SH SYNOPSIS
.B omorfi-vislcg.bash
[\fI\,OPTION\/\fR] [\fI\,FILENAME\/\fR...]
.SH DESCRIPTION
Analyse text and output undisambiguated VISL CG 3 format.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print this help dialog
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-X\fR, \fB\-\-all\-words\fR
Use large\-coverage dictionary
.PP
If no FILENAMEs are given, input is read from standard input.
.SH EXAMPLE
Analyse text file:
.PP
omorfi-vislcg.bash test.text
.SH COPYRIGHT
Copyright \(co 2018 Tommi A Pirinen
Licence GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
cg-proc(1)


NEWS
# NEWS

This file lists noteworthy changes between releases, for full list of changes,
see git log and then `ChangeLog.old`.

## Significant changes in 20180511

* Universal dependencies version 2.2 is now used as target
* At least 226 new words: 239 additions and 13 deletions in lexeme database
* Most changes are in development infra, so not visible to end users...
* Started rewriting CG from the scratch
* The APIs for programming language deprecate load(filename) and load(dir) forms
  of filename guessing functions in favour of forthcoming loadAnalyser(file),
  loadLemmatiser(file), loadUDPipe(file) etc. etc. functions
* Working towards more general tokenise-analyse-disambiguate pipelines maybe,
  or just refactoring
* lots more automated tests -> lots less human errors
* By popular request: there are two analysers now, one with small dictionary
  and one with full, use the smaller one when you do not want to see birds or
  languages or tribes analysed. The smaller one replaces the old default, but
  the new tools will require you to select one explicitly anyways
* fixes and workarounds: java and c++ can now be disabled partially or totally
* adopted SG0 as possible verb form analysis from UD data
* The end users are now provided with bash-scripts wrappers for all
  functionalities, whereas the typically python versions allow more control
  of parametres


## Significant changes in 20170515

* Universal Dependencies version 2 is now used, still mainly lemma, UPOS,
  features fields are analysed
* At least 2,336 new words (based on diffstat: 38886 additions, 3655 deletions)
* Preliminary support for various guessing models: python-based, finite-state
  and UDPipe. This means that it is possibly to get analyses for all tokens,
  albeit quality of guesses varies.
* A minimal C++ library version has been made to match java and python bindings.
  C++-11 and libhfst are required.
* The dix version can now be compiled with lttoolbox with a lot of memory
* A restricted "gold" dictionary mode has been added. This is good for both end
  users with limited memory and end users who require higher quality lexemes
  (i.e., only research institute approved, no wiktionary words or other weird
   stuffs)
* Documentations and automatic testing much reworked with the new modern toys
  from github: travis-ci, jekyll
* Started weeding the ADP/ADV jungle...
* Fixed a horrible bug in the corpus coverage testing that terribly
  under-estimated our coverage for corpora where hapax legomena etc. were
  ignored
* Lot of documentation has been semi-automated, therefore many changes can be
  viewed at the new gh-pages site: https://flammie.github.io/omorfi/

## Significant changes in 20161115

* Started drafting more blacklists and *known good* lexemes subsets for people
  who struggle with rare words and productive compounding, derivation
* Updated to Universal Dependencies version 1.4
* A lot of new derivations by the way
* Preliminary guessers
* More loopy guessery things for punctuation and digit combos
* Minor fixes to UD feature sorting
* Homonym numbers used in some applications
* Added timeouts where downstream tools support them, so tools don't seem like
  they are freezing at random
* moved old documentations to github-pages
* added preliminary hfst-pmatch-based tokeniser

## Significant changes in 20160515

* Universal Dependencies for Finnish is the new standard format we now follow:
    * POS is now UPOS and classes were changed accordingly (new classes: AUX,
      PROPN, DET, CONJ, SCONJ, PUNCT, SYM, and VERB, NOUN, ADP, ADV as before)
    * other features mostly match the feature field in UD documentation
    * release cycle aims to be same six month cycle as with UD
    * the automatic tests verify compatibility with UD; 92 % of lemmas, primary
      POS tags and morphological features are the same as Finnish UD corpus,
      75 % same as Finnish FTB UD corpus
    * analyser for reading and writing CONLL-U format
* tokenisation as script and more hacks to token stripping in corner cases
* continuous integration with travis-ci, currently only testing basic script
  programming conventions
* added a lot of high coverage words and forms by hand
* by popular request, some of the words can now be blacklisted, when you don't
  want that guy named Mutta to ambiguate your conjunction analyses or the odd
  new guinean bird to clash with some common verb
* the "database" is now only keyed on lemma + homonym number; paradigm is extra
  information like anything else
* a lot of work on morphological segmentation towards statistical machine
  translation; check proceedings of WMT shared tasks 2015 and 2016 to see why
* started refactoring some python code into classes

## Significant changes in 20150904

* allomorphy can be tagged again to distinguish e.g. *-iden* and *-itten* when
  generating
* FinnTreeBank-1 format provided by Miikka Silfverberg is available but not
  built by default since it lacks a test set
* lexicalised inflections can have separate tag, e.g. *kännissä* can be lexical
  inessive distinguished from regular inessive
* preliminary VISL CG-3 support, with original grammar by Fred Karlsson;
  convenience bash scripts available for disambiguated parsing
* preliminary support for conllu and conllx analysis formats
* paradigm categorisation is now verified by regular expressions
* lots of paradigm fixes and some added words

## Significant changes in 20150326

* speed is up to >20,000 tokens per second from ~500
* coverages are up to:
  europarl (99 %) gutenberg (97 %), JRC Acquis (94 %) and fiwiki (93 %)
* moses factored model format supported
* segmentation supported
* Java API
* Python hacks packaged to API and module
* Rest of hand-written Xerox legacy data removed; all is script-generated
* github migration since google code is EOL'd
* file naming for automata changed to include omorfi prefix for all file
  names in case they are distributed separately.

## Significant changes in 20141014


src/python/omorfi-vislcg.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# string munging
import re
from argparse import ArgumentParser, FileType
# CLI stuff
from sys import stdin, stdout, stderr
# statistics
from time import perf_counter, process_time

# omorfi
from omorfi.omorfi import Omorfi
from omorfi.token import get_lemmas, \
        get_vislcg_feats, get_line_tokens, get_line_tokens_vislcg, \
        get_line_tokens_conllu, format_analyses_vislcg


def print_analyses_vislcg3(surf, anals, outfile):
    print(format_analyses_vislcg(surf, anals), file=outfile)

def main():
    """Invoke a simple CLI analyser."""
    a = ArgumentParser()
    a.add_argument('-a', '--analyser', metavar='AFILE',
                   help="load analyser model from AFILE")
    a.add_argument('-i', '--input', metavar="INFILE", type=open,
                   dest="infile", help="source of analysis data")
    a.add_argument('-v', '--verbose', action='store_true',
                   help="print verbosely while processing")
    a.add_argument('-o', '--output', metavar="OUTFILE", dest="outfile",
                   help="print output into OUTFILE", type=FileType('w'))
    a.add_argument('-F', '--format', metavar="INFORMAT", default='text',
                   help="read input using INFORMAT tokenisation",
                   choices=['text', 'vislcg', 'conllu'])
    a.add_argument('-x', '--statistics', metavar="STATFILE", dest="statfile",
                   help="print statistics to STATFILE", type=FileType('w'))
    options = a.parse_args()
    omorfi = Omorfi(options.verbose)
    if options.analyser:
        if options.verbose:
            print("reading analyser model", options.analyser)
        omorfi.load_analyser(options.analyser)
    else:
        print("analyser is required to vislcg", file=stderr)
        exit(4)
    if not options.infile:
        options.infile = stdin
    if options.verbose:
        print("analysing", options.infile.name)
    if not options.outfile:
        options.outfile = stdout
    if options.verbose:
        print("writing to", options.outfile.name)
    if not options.statfile:
        if options.outfile == stdout:
            options.statfile = stdout
        else:
            options.statfile = stderr
    # statistics
    realstart = perf_counter()
    cpustart = process_time()
    tokens = 0
    unknowns = 0
    last = None
    for line in options.infile:
        surfs = []
        if options.format == 'vislcg':
            surfs = get_line_tokens_vislcg(line, last)
        elif options.format == 'text':
            surfs = get_line_tokens(line, omorfi)
        elif options.format == 'conllu':
            surfs = get_line_tokens_conllu(line, last)
        else:
            print("input format missing implementation", options.format,
                  file=stderr)
            exit(2)
        for surf in surfs:
            if 'conllu_form' in surf:
                # skip conllu special forms in input for now:
                # (ellipsis and MWE magics)
                continue
            elif 'surf' in surf:
                tokens += 1
                anals = omorfi.analyse(surf)
                if len(anals) == 0 or (len(anals) == 1 and
                                       'UNKNOWN' in anals[0]['anal']):
                    unknowns += 1
                    anals = omorfi.guess(surf)
                print_analyses_vislcg3(surf, anals, options.outfile)
            elif 'comment' in surf:
                if surf['comment'].startswith(';') or \
                       surf['comment'].startswith('\t'):
                    continue
                else:
                    print(surf['comment'], file=options.outfile)
            elif 'error' in surf:
                print(surf['error'], file=stderr)
                exit(2)
            last = surf
    cpuend = process_time()
    realend = perf_counter()
    print("# Tokens:", tokens, "\n# Unknown:", unknowns,
          unknowns / tokens * 100, "%", file=options.statfile)
    print("# CPU time:", cpuend - cpustart,
          "\n# Real time:", realend - realstart,
          file=options.statfile)
    print("# Tokens per timeunit:", tokens / (realend - realstart),
          file=options.statfile)
    exit(0)


if __name__ == "__main__":
    main()
